dialogue-core/src/main/java/com/palantir/dialogue/core/LockFreeTimeWindowReservoir.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.github.benmanes.caffeine.cache.Ticker;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Stores marks on a timeline of buckets, where marks are forgotten after the specified {@code memory}.
+ * Reads and writes are lock-free and efficient.
+ *
+ * Replacement for {@link com.codahale.metrics.SlidingTimeWindowArrayReservoir}.
+ */
+@ThreadSafe
+final class LockFreeTimeWindowReservoir {
+    private final long[] buckets;
+    private final long bucketSizeNanos;
+    private final Ticker clock;
+
+    /** Cursor tells us which bucket to write marks into. Always less than buckets.length, always advances. */
+    private final AtomicInteger cursor = new AtomicInteger(0);
+    /** We maintain an extra count variable to serve fast reads - it should always equal the sum of the buckets. */
+    private final AtomicLong count = new AtomicLong(0);
+    /** Time in nanos when we need to move the cursor to the next bucket (and expire its contents). */
+    private final AtomicLong nextRollover;
+
+    LockFreeTimeWindowReservoir(Duration memory, int granularity, Ticker clock) {
+        this.buckets = new long[granularity];
+        this.clock = clock;
+        this.bucketSizeNanos = memory.toNanos() / granularity;
+        this.nextRollover = new AtomicLong(clock.read() + bucketSizeNanos);
+    }
+
+    void mark() {
+        maybeRoll();
+        buckets[cursor.get()] += 1;
```

Comment:
buckets probably needs to be a `AtomicLong[]` so this behaves under concurrency, right?